### PR TITLE
Implementation of TrimZero and BGRU.

### PR DESCRIPTION
**Pull request by jnhwkim**
`TrimZero` has the same result of `MaskZero`'s, however, `TrimZero` is faster than `MaskZero` only when sentence lengths is costly vary. In practice, e.g. language model, `TrimZero` is expected to be faster than `MaskZero` about 30%. (You can test with it using `test/test_trimzero.lua`.)

`BGRU` is implementation of RnnDrop (Moon et al., 2015) or Bayesian RNN (Gal, 2015), a novel method to dropout on inner connections of RNNs.

### CODE REVIEW

File: Dropout.lua [hunk 1]
```diff
@@ -0,0 +1,70 @@
+local Dropout, Parent = nn.Dropout, nn.Module
```
Comment (nicholas-leonard) [inline Dropout.lua]: Why do we redefine Dropout? It is already in nn right?

File: TrimZero.lua [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      
+      -- make sure mask has the same dimenion as the input tensor
+      local inputSize = input:size():fill(1)
+      if input:dim() - 1 == self.nInputDim then
+         inputSize[1] = input:size(1)
+      end
+      mask:resize(inputSize)
+      -- build mask
+      if input:dim() - 1 == self.nInputDim then
+         output:resizeAs(input):copy(input)
```
Comment (nicholas-leonard) [inline TrimZero.lua]: Could you use a lazily allocated buffer here to reduce memory allocations in the following `output = output:index...`. So instead we would have `output:index(buffer, ....`

File: TrimZero.lua [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+         assert(torch.find, 'install torchx package : luarocks install torchx')
+         local indexes = torch.find(mask, 0)
+         if 0 < #indexes then
+            output = output:index(1, torch.LongTensor(indexes))
+         else
+            output = output:index(1, torch.LongTensor{1}):zero()
+         end
+      else
+         if mask[1] == 1 then output = input:clone():zero()
```
Comment (nicholas-leonard) [inline TrimZero.lua]: `output:resize(input:size()):zero()` will reuse existing memory.

File: Dropout.lua (hunk 1 above)
Comment (jnhwkim) [inline Dropout.lua]: @nicholas-leonard I added `lazy` option to only resample after backward is called. This mechanism is used by Bayesian GRUs to use the same dropout mask for each sequence, not for each word. See [A Theoretically Grounded Application of Dropout in Recurrent Neural Networks by Gal (2015)](http://arxiv.org/abs/1512.05287).

File: TrimZero.lua (hunk 3 above)
Comment (jnhwkim) [inline TrimZero.lua]: @nicholas-leonard I got it. thanks!

File: Dropout.lua (hunk 1 above)
Comment (nicholas-leonard) [inline Dropout.lua]: @jnhwkim Okay that is awesome. Can we include that bit of doc in the header of the dropout file? Thanks.

File: TrimZero.lua (hunk 2 above)
Comment (jnhwkim) [inline TrimZero.lua]: @nicholas-leonard Could you check my reimplementation? I'll update this part in a little bit different way.

File: Dropout.lua (hunk 1 above)
Comment (jnhwkim) [inline Dropout.lua]: done!

**Comment by jnhwkim** [comment]
@nicholas-leonard Updated. Can you check if I miss?

File: Dropout.lua (hunk 1 above)
Comment (jnhwkim) [inline Dropout.lua]: @nicholas-leonard added.

Comment (nicholas-leonard) [inline Dropout.lua]: @jnhwkim Looks good.

**Comment by jnhwkim** [comment]
@nicholas-leonard recent commits are merged, a conflict about `examples/recurrent-language-model.lua` is resolved in a conservative way while applying a new dropout technique on GRUs (BGRU).

**Comment by nicholas-leonard** [comment]
@jnhwkim Awesome PR. Thanks!

**Comment by jnhwkim** [comment]
@nicholas-leonard You're welcome!
